oauth2_http/java/com/google/auth/oauth2/GdchCredentials.java

```diff
@@ -0,0 +1,509 @@
+    this.lifetime = builder.lifetime;
+  }
+
+  static class TransportFactoryForGdch implements HttpTransportFactory {
+    HttpTransport transport;
+
+    public TransportFactoryForGdch(String caCertPath) throws IOException {
+      setTransport(caCertPath);
+    }
+
+    @Override
+    public HttpTransport create() {
+      return transport;
+    }
+
+    private void setTransport(String caCertPath) throws IOException {
+      try {
+        InputStream certificateStream = readStream(new File(caCertPath));
+        this.transport =
+            new NetHttpTransport.Builder().trustCertificatesFromStream(certificateStream).build();
+      } catch (IOException e) {
+        throw new IOException(
+            String.format(
+                "Error reading certificate file from CA cert path, value '%s': %s",
+                caCertPath, e.getMessage()),
+            e);
+      } catch (GeneralSecurityException e) {
+        throw new IOException("Error initiating transport with certificate stream.", e);
+      }
+    }
+  }
+
+  /**
+   * Create GDCH service account credentials defined by JSON.
+   *
+   * @param json a map from the JSON representing the credentials.
+   * @param transportFactory HTTP transport factory, creates the transport used to get access
+   *     tokens.
+   * @return the GDCH service account credentials defined by the JSON.
+   * @throws IOException if the credential cannot be created from the JSON.
+   */
+  static GdchCredentials fromJson(Map<String, Object> json, HttpTransportFactory transportFactory)
+      throws IOException {
+    String formatVersion = validateField((String) json.get("format_version"), "format_version");
+    String projectId = validateField((String) json.get("project"), "project");
+    String privateKeyId = validateField((String) json.get("private_key_id"), "private_key_id");
+    String privateKeyPkcs8 = validateField((String) json.get("private_key"), "private_key");
+    String serviceIdentityName = validateField((String) json.get("name"), "name");
+    String tokenServerUriStringFromCreds =
+        validateField((String) json.get("token_uri"), "token_uri");
+
+    /**
+     * The CA cert path for token server side TLS, certificate verification. If the token server
+     * uses well known CA, then this parameter can be `None`.
+     */
+    String caCertPath = (String) json.get("ca_cert_path");
+
+    if (!SUPPORTED_FORMAT_VERSION.equals(formatVersion)) {
+      throw new IOException(
+          String.format("Only format version %s is supported.", SUPPORTED_FORMAT_VERSION));
+    }
+
+    URI tokenServerUriFromCreds = null;
+    try {
+      tokenServerUriFromCreds = new URI(tokenServerUriStringFromCreds);
+    } catch (URISyntaxException e) {
+      throw new IOException("Token server URI specified in 'token_uri' could not be parsed.");
+    }
+
+    // Override the transportFactory if CA cert path is provided.
+    if (caCertPath != null && caCertPath.length() > 0) {
```

Comment:
Should we use default the httptransport if the ca cert path is not provided?

Reply:
yes

Reply:
it is required, we throw otherwise